From 702aa37d6ae67432b8c0bd190e7f428682d3c345 Mon Sep 17 00:00:00 2001
From: Sugar Zhang <sugar.zhang@rock-chips.com>
Date: Mon, 22 Nov 2021 14:22:53 +0800
Subject: [PATCH 83/94] drm/bridge: synopsys: Add bitstream support for hdmi-qp

Signed-off-by: Sugar Zhang <sugar.zhang@rock-chips.com>
Change-Id: I83e5558507994351b182a503325a56507f1867ee
Signed-off-by: Detlev Casanova <detlev.casanova@collabora.com>

Gbp-Pq: Name 0164-drm-bridge-synopsys-Add-bitstream-support-for-hdmi-q.patch
---
 .../bridge/synopsys/dw-hdmi-qp-i2s-audio.c    | 19 +++++++++++++++----
 drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c  |  5 +++++
 2 files changed, 20 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp-i2s-audio.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp-i2s-audio.c
index c00712666..464e35cc4 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp-i2s-audio.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp-i2s-audio.c
@@ -88,11 +88,22 @@ static int dw_hdmi_qp_i2s_hw_params(struct device *dev, void *data,
 
 	hdmi_mod(audio, conf0, I2S_LINES_EN_MSK, AUDIO_INTERFACE_CONFIG0);
 
-	/* Enable bpcuv generated internally */
-	hdmi_mod(audio, I2S_BPCUV_RCV_DIS, I2S_BPCUV_RCV_MSK, AUDIO_INTERFACE_CONFIG0);
+	/*
+	 * Enable bpcuv generated internally for L-PCM, or received
+	 * from stream for NLPCM/HBR.
+	 */
+	switch (fmt->bit_fmt) {
+	case SNDRV_PCM_FORMAT_IEC958_SUBFRAME_LE:
+		conf0 = (hparms->channels == 8) ? AUD_HBR : AUD_ASP;
+		conf0 |= I2S_BPCUV_RCV_EN;
+		break;
+	default:
+		conf0 = AUD_ASP | I2S_BPCUV_RCV_DIS;
+		break;
+	}
 
-	/* Configure the audio format */
-	hdmi_mod(audio, AUD_ASP, AUD_FORMAT_MSK, AUDIO_INTERFACE_CONFIG0);
+	hdmi_mod(audio, conf0, I2S_BPCUV_RCV_MSK | AUD_FORMAT_MSK,
+		 AUDIO_INTERFACE_CONFIG0);
 
 	/* Enable audio FIFO auto clear when overflow */
 	hdmi_mod(audio, AUD_FIFO_INIT_ON_OVF_EN, AUD_FIFO_INIT_ON_OVF_MSK,
diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
index 10a90527e..89e35948f 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
@@ -280,6 +280,11 @@ static unsigned int hdmi_find_n(struct dw_hdmi_qp *hdmi, unsigned long pixel_clk
 void dw_hdmi_qp_set_channel_status(struct dw_hdmi_qp *hdmi,
 				   u8 *channel_status)
 {
+	/* Set channel status */
+	hdmi_writel(hdmi, channel_status[3] | (channel_status[4] << 8),
+		    AUDPKT_CHSTATUS_OVR1);
+	hdmi_modb(hdmi, AUDPKT_CHSTATUS_OVR_EN,
+		  AUDPKT_CHSTATUS_OVR_EN_MASK, AUDPKT_CONTROL0);
 }
 EXPORT_SYMBOL_GPL(dw_hdmi_qp_set_channel_status);
 
-- 
2.46.0

